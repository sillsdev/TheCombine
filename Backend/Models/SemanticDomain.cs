using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.Linq;

namespace BackendFramework.Models
{
    public class SemanticDomain
    {
        [Required]
        public string Name { get; set; }
        [Required]
        public string Id { get; set; }

        public SemanticDomain()
        {
            Name = "";
            Id = "";
        }

        public SemanticDomain Clone()
        {
            return new SemanticDomain
            {
                Name = (string)Name.Clone(),
                Id = (string)Id.Clone()
            };
        }

        public override bool Equals(object? obj)
        {
            if (obj is not SemanticDomain other || GetType() != obj.GetType())
            {
                return false;
            }

            return Name.Equals(other.Name) && Id.Equals(other.Id);
        }

        public override int GetHashCode()
        {
            return HashCode.Combine(Name, Id);
        }
    }

    public class SemanticDomainFull
    {
        [Required]
        public string Name { get; set; }
        [Required]
        public string Id { get; set; }
        [Required]
        public string Description { get; set; }
        [Required]
        public List<string> Questions { get; set; }

        public SemanticDomainFull()
        {
            Name = "";
            Id = "";
            Description = "";
            Questions = new List<string>();
        }

        public SemanticDomainFull Clone()
        {
            var clone = new SemanticDomainFull
            {
                Name = (string)Name.Clone(),
                Id = (string)Id.Clone(),
                Description = (string)Description.Clone(),
                Questions = new List<string>()
            };

            foreach (var question in Questions)
            {
                clone.Questions.Add((string)question.Clone());
            }

            return clone;
        }

        public override bool Equals(object? obj)
        {
            if (obj is not SemanticDomainFull other || GetType() != obj.GetType())
            {
                return false;
            }

            return
                Name.Equals(other.Name) &&
                Id.Equals(other.Id) &&
                Description.Equals(other.Description) &&
                Questions.Count == other.Questions.Count &&
                Questions.All(other.Questions.Contains);
        }

        public override int GetHashCode()
        {
            return HashCode.Combine(Name, Id, Description, Questions);
        }
    }

    public class SemanticDomainTreeNode
    {
        [Required]
        public SemanticDomain Node { get; set; }
        [Required]
        public SemanticDomain Parent { get; set; }
        [Required]
        public SemanticDomain Previous { get; set; }
        [Required]
        public SemanticDomain Next { get; set; }
        [Required]
        public List<SemanticDomain> Children { get; set; }

        public SemanticDomainTreeNode()
        {
            Node = new SemanticDomain();
            Parent = new SemanticDomain();
            Previous = new SemanticDomain();
            Next = new SemanticDomain();
            Children = new List<SemanticDomain>();
        }

        public SemanticDomainTreeNode Clone()
        {
            var clone = new SemanticDomainTreeNode
            {
                Node = Node.Clone(),
                Parent = Parent.Clone(),
                Previous = Previous.Clone(),
                Next = Next.Clone(),
                Children = new List<SemanticDomain>()
            };

            foreach (var child in Children)
            {
                clone.Children.Add(child.Clone());
            }

            return clone;
        }

        public override bool Equals(object? obj)
        {
            if (obj is not SemanticDomainTreeNode other || GetType() != obj.GetType())
            {
                return false;
            }

            return
                Node.Equals(other.Node) &&
                Parent.Equals(other.Parent) &&
                Previous.Equals(other.Previous) &&
                Next.Equals(other.Next) &&
                Children.Count == other.Children.Count &&
                Children.All(other.Children.Contains);
        }

        public override int GetHashCode()
        {
            return HashCode.Combine(Node, Parent, Previous, Next, Children);
        }
    }

    /// <remarks>
    /// This is used in an OpenAPI return value serializer, so its attributes must be defined as properties.
    /// </remarks>
    public class SemanticDomainWithSubdomains
    {
        [Required]
        public string Name { get; set; }
        [Required]
        public string Id { get; set; }
        [Required]
        public string Description { get; set; }
        [Required]
        public List<SemanticDomainWithSubdomains> Subdomains { get; set; }

        public SemanticDomainWithSubdomains(SemanticDomain sd)
        {
            Name = sd.Name;
            Id = sd.Id;
            Description = "";
            Subdomains = new List<SemanticDomainWithSubdomains>();
        }
    }
}
