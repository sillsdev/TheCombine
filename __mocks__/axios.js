// This file is a manual axios mock. For an explanation, see the mockExplanation document
// on docs

export default {
  // get: when axios.get is called within parent directory, instead call this function
  // The data stored here should be moved to a directory called __fixtures__.
  // That will come in a later pull request.
  get: jest.fn(route => {
    return Promise.resolve({
      data: {
        id: "1",
        edits: []
      }
    });
  }),

  // post: when axios.post is called within parent directory, instead call this function
  post: jest.fn(route => {
    return Promise.resolve({
      types: "post"
    });
  }),

  // create: when axios.create is called within parent directory, instead return this object.
  // mockReturnThis() is just shorthand for jest.fn(function() { return this; });
  create: jest.fn().mockReturnThis()
};
