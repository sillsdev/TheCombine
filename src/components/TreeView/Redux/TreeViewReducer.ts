import {
  TreeViewAction,
  TreeActionType,
  TreeViewState,
  defaultState,
} from "components/TreeView/Redux/TreeViewReduxTypes";
import { StoreAction, StoreActionTypes } from "rootActions";

export const treeViewReducer = (
  state: TreeViewState = defaultState,
  action: StoreAction | TreeViewAction
): TreeViewState => {
  switch (action.type) {
    case TreeActionType.CLOSE_TREE:
      return { ...state, open: false };
    case TreeActionType.OPEN_TREE:
      return { ...state, open: true };
    case TreeActionType.RESET_TREE:
      return defaultState;
    case TreeActionType.SET_DOMAIN_LANGUAGE:
      if (!action.language) {
        throw new Error("Cannot set domain language to undefined.");
      }
      return {
        ...state,
        currentDomain: { ...state.currentDomain, lang: action.language },
        language: action.language,
      };
    case TreeActionType.SET_CURRENT_DOMAIN:
      if (!action.domain) {
        throw new Error("Cannot set the current domain to undefined.");
      }
      return { ...state, currentDomain: action.domain };
    case StoreActionTypes.RESET:
      return defaultState;
    default:
      return state;
  }
};
