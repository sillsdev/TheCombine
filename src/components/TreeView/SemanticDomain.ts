export default interface SemanticDomainWithSubdomains {
  // Data of current domain
  name: string;
  id: string;

  // Data about tangential domains
  parentDomain?: SemanticDomainWithSubdomains;
  subdomains: SemanticDomainWithSubdomains[];
}
