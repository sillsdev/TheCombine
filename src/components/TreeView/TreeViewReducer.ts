import { TreeViewAction, TreeActionType } from "./TreeViewActions";
import SemanticDomainWithSubdomains from "./SemanticDomain";
import { StoreAction, StoreActions } from "../../rootActions";

export interface TreeViewState {
  currentdomain: SemanticDomainWithSubdomains;
}

// Parses a list of semantic domains (to be received from backend)
export function createDomains(
  data: SemanticDomainWithSubdomains[]
): TreeViewState {
  let state: TreeViewState = {
    currentdomain: {
      ...defaultState.currentdomain,
      subdomains: data
    }
  };
  addParentDomains(state.currentdomain);
  // while (state.currentDomain.subDomains.length > 0)
  //   state.currentDomain = state.currentDomain.subDomains[0];
  return state;
}

// Adds the parent domains to the information sent by the backend
function addParentDomains(parent: SemanticDomainWithSubdomains) {
  if (parent.subdomains)
    for (let domain of parent.subdomains) {
      domain.parentDomain = parent;
      addParentDomains(domain);
    }
  //else parent.subDomains = [];
}

// Creates a dummy default state
export const defaultState: TreeViewState = {
  currentdomain: {
    name: "",
    id: "",
    description: "",
    subdomains: []
  }
};

export const treeViewReducer = (
  state: TreeViewState = defaultState,
  action: StoreAction | TreeViewAction
): TreeViewState => {
  switch (action.type) {
    case TreeActionType.TRAVERSE_TREE:
      return { ...state, currentdomain: action.payload };
    case StoreActions.RESET:
      return defaultState;
    default:
      return state;
  }
};
