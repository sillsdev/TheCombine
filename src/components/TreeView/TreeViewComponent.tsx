import React from "react";
import { Grid, Zoom } from "@material-ui/core";

import TreeProps from "./TreeProps";
import TreeDepiction from "./TreeDepiction";
import SemanticDomainWithSubdomains from "./SemanticDomain";
import { createDomains } from "./TreeViewReducer";

// Domain data
import en from "../../resources/semantic-domains/en.json";

interface TreeViewProps extends TreeProps {
  returnControlToCaller: () => void;
}

interface TreeViewComponentState {
  visible: boolean;
}

export default class TreeViewComponent extends React.Component<
  TreeViewProps,
  TreeViewComponentState
> {
  constructor(props: TreeViewProps) {
    super(props);
    this.state = { visible: true };

    this.animate = this.animate.bind(this);

    // TODO: add checking the user's language to select the semantic domains
    let newDomain = createDomains(en);
    this.props.navigate(newDomain.currentdomain);
  }

  animate(domain: SemanticDomainWithSubdomains | undefined): Promise<void> {
    if (this.state.visible) {
      this.setState({ visible: false });
      return new Promise(resolve =>
        setTimeout(() => {
          if (domain && this.state.visible === false) {
            if (domain.id !== this.props.currentDomain.id) {
              this.props.navigate(domain);
              this.setState({ visible: true });
            } else {
              this.props.returnControlToCaller();
            }
          }
          resolve();
        }, 300)
      );
    } else return Promise.reject("Change already in-progress");
  }

  render() {
    return (
      <React.Fragment>
        <Zoom in={this.state.visible}>
          <Grid
            container
            direction="column"
            justify="center"
            alignItems="center"
          >
            <TreeDepiction
              currentDomain={this.props.currentDomain}
              animate={this.animate}
            />
          </Grid>
        </Zoom>
      </React.Fragment>
    );
  }
}
