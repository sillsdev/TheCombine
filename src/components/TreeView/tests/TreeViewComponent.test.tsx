import React from "react";
import renderer, { ReactTestRenderer } from "react-test-renderer";
import TreeViewComponent from "../TreeViewComponent";
import { defaultState } from "../TreeViewReducer";
import SemanticDomain from "../SemanticDomain";
import axios from "axios";
import mockTree from "./MockSemanticTree";

var treeMaster: ReactTestRenderer;
var treeHandle: TreeViewComponent;

const RETURN_MOCK = jest.fn();
const NAVIGATE_MOCK = jest.fn();

// Mock out setTimeout
jest.useFakeTimers();

// Mock out Zoom to avoid issues with portals
jest.mock("@material-ui/core", () => {
  const realMaterialUi = jest.requireActual("@material-ui/core");
  return {
    ...realMaterialUi,
    Zoom: realMaterialUi.Container
  };
});

// Mock axios instance to mock return data
const mockAxios = axios as jest.Mocked<typeof axios>;

beforeEach(() => {
  mockAxios.get.mockImplementationOnce(() => Promise.resolve({ data: [defaultState.currentdomain] }));
  renderer.act(() => {
    treeMaster = renderer.create(
      <TreeViewComponent
        currentDomain={mockTree}
        returnControlToCaller={RETURN_MOCK}
        navigate={NAVIGATE_MOCK}
      />
    );
  });
  treeHandle = treeMaster.root.findByType(TreeViewComponent).instance;

  RETURN_MOCK.mockClear();
  NAVIGATE_MOCK.mockClear();
});

describe("Tests AddWords", () => {
  it("Constructs correctly", () => {
    // Default snapshot test
    snapTest("default view");
    expect(NAVIGATE_MOCK).toHaveBeenCalledWith({
      name: "Semantic Domains",
      id: "", subdomains: [defaultState.currentdomain]});
  });

  it("Sets a new domain upon navigation", () => {
    let newDom: SemanticDomain = {
      name: "test",
      id: "test",
      subdomains: []
    };

    treeHandle.animate(newDom);
    jest.runAllTimers();

    expect(NAVIGATE_MOCK).toHaveBeenCalledWith(newDom);
  });

  it("Returns control to caller when the same semantic domain is passed in", () => {
    treeHandle.animate(mockTree);
    jest.runAllTimers();

    expect(RETURN_MOCK).toHaveBeenCalledTimes(1);
  });
});

// Perform a snapshot test
function snapTest(name: string) {
  expect(treeMaster.toJSON()).toMatchSnapshot();
}
