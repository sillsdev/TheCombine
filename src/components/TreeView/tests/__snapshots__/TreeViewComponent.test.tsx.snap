// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests AddWords Renders correctly 1`] = `
<div
  className="MuiContainer-root MuiContainer-maxWidthLg"
  in={true}
>
  <div
    className="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-column MuiGrid-align-items-xs-center MuiGrid-justify-xs-center"
  >
    <div
      className="MuiGrid-root MuiGrid-item"
    />
    <div
      className="MuiGrid-root MuiGrid-item"
    >
      <ul
        className="MuiGridList-root"
        style={
          Object {
            "margin": -1,
          }
        }
      >
        <li
          className="MuiGridListTile-root"
          style={
            Object {
              "height": "auto",
              "padding": 1,
              "width": "20%",
            }
          }
        >
          <div
            className="MuiGridListTile-tile"
          />
        </li>
        <li
          className="MuiGridListTile-root"
          style={
            Object {
              "height": "auto",
              "padding": 1,
              "width": "60%",
            }
          }
        >
          <div
            className="MuiGridListTile-tile"
          >
            <div
              className="MuiPaper-root MuiPaper-elevation1 MuiCard-root MuiPaper-rounded"
            >
              <button
                className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeLarge MuiButton-fullWidth"
                disabled={false}
                onBlur={[Function]}
                onClick={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                onMouseDown={[Function]}
                onMouseLeave={[Function]}
                onMouseUp={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
                tabIndex={0}
                type="button"
              >
                <span
                  className="MuiButton-label"
                >
                  <div
                    style={
                      Object {
                        "textTransform": "capitalize",
                      }
                    }
                  >
                    <span
                      className="MuiTypography-root MuiTypography-overline"
                    >
                      1.1.1.1
                    </span>
                    <h6
                      className="MuiTypography-root MuiTypography-h6"
                    >
                      moon
                    </h6>
                  </div>
                </span>
                <span
                  className="MuiTouchRipple-root"
                />
              </button>
              <div
                className="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal MuiFormControl-fullWidth"
                onKeyUp={[Function]}
              >
                <label
                  className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
                  data-shrink={false}
                  htmlFor="name"
                >
                  Find a domain
                </label>
                <div
                  className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
                  onClick={[Function]}
                >
                  <input
                    aria-invalid={false}
                    autoComplete="off"
                    className="MuiInputBase-input MuiInput-input"
                    disabled={false}
                    id="name"
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    required={false}
                    type="text"
                  />
                </div>
              </div>
            </div>
          </div>
        </li>
        <li
          className="MuiGridListTile-root"
          style={
            Object {
              "height": "auto",
              "padding": 1,
              "width": "20%",
            }
          }
        >
          <div
            className="MuiGridListTile-tile"
          />
        </li>
      </ul>
    </div>
    <div
      className="MuiGrid-root MuiGrid-item"
    >
      <ul
        className="MuiGridList-root"
        style={
          Object {
            "margin": -0,
            "width": 150,
          }
        }
      >
        <li
          className="MuiGridListTile-root"
          style={
            Object {
              "height": "auto",
              "padding": 0,
              "width": "100%",
            }
          }
        >
          <div
            className="MuiGridListTile-tile"
          >
            <img
              alt="parent.svg"
              height="100%"
              src="parent.svg"
              width={150}
            />
          </div>
        </li>
        <li
          className="MuiGridListTile-root"
          style={
            Object {
              "height": "auto",
              "padding": 0,
              "width": "100%",
            }
          }
        >
          <div
            className="MuiGridListTile-tile"
          >
            <button
              className="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary"
              disabled={false}
              id="1.1.1"
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              style={
                Object {
                  "bottom": 0,
                  "height": "90%",
                  "left": 0,
                  "margin": "5%",
                  "width": "90%",
                }
              }
              tabIndex={0}
              type="button"
            >
              <span
                className="MuiButton-label"
              >
                <div
                  style={
                    Object {
                      "textTransform": "capitalize",
                    }
                  }
                >
                  <span
                    className="MuiTypography-root MuiTypography-overline"
                  >
                    1.1.1
                  </span>
                  <p
                    className="MuiTypography-root MuiTypography-body1"
                  >
                    sun
                  </p>
                </div>
              </span>
              <span
                className="MuiTouchRipple-root"
              />
            </button>
          </div>
        </li>
      </ul>
    </div>
  </div>
</div>
`;
